### PR TITLE
Customizable URLs for studio

### LINKED ISSUE #1336: Customizable URLs
We would like to allow OCW Studio users to be able to customize site URL and helping them to apply some consistent url patterns for all courses/websites.

## 1. Allow content authors in OCW Studio  to set a custom path for their websites

* The url of the website should be set by the user when they  publish the site for the first time. By default it should be the slugified version of the `Website.name` (unless the starter config has a defined `site-url-format` attribute as described in # 2 below).

00_default

* Once the site is published, the path cannot be edited by users in Studio.
* At this point the url of the website is saved in the database (in a new `website.url` field)
* url must be unique, obviously. 

## 2. Suggest a consistent url pattern for all websites of a certain type (e.g. courses)

* Website starters can suggest a certain url pattern for all website that use that specific starter. Something like

`url-format : [metadata:primary_course_nr]-[metadata:course_title]-[metadata:term]-[metadata-year] `

Slugified version of each field is assumed (periods converted to dashes etc.)

* By default the publishing drawer would look something like 

cutom-url - 01-starter-defautl

That means the user has not populated any of the relevant fields in the metadata. 

* If they have populated one of the fields, e.g. year, then it will looks something like 

cutom-url - 02-starter-one

* The website cannot be published unless all fields in the url are resolved, which can happen by a) user populating all metadata fields or b) user decides to use their own url (overrides the suggested url format from the starter config

### PR DESCRIPTION
**Pull request by mbertrand**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
Closes #1336

#### What's this PR do?
- Adds a `url_path` field to `Website`.  The migration should prepopulate it with the same value as `Website.name` for all sites that have already been published to production.
- When publishing to draft/live, `url_path` can be sent in the request body to set the field's value.  The serializer should allow the field value to change as long as the site hasn't been published to production yet and the value is valid.  If the value is changed, all `WebsiteContent` objects with populated `file` values for the site will have their synced_checksums reset so they will be resynced to github with the updated path.
- Updates the HugoMarkdownSerializer to save the correct `file` values (based on `url_path`) to the backend (git) and vice versa.
- Uses a new `site-url-format` field in site config to suggest a pattern for this field to the user on the front end.
- Changes the publish drawer form, to include a url field if the site has not been published to production yet.
- Updates both individual and mass pipeline definitions so that S3 files are copied to the right places.  Creates/updates the site pipeline before each publish event to ensure that it has the correct url path baked in.

#### How should this be manually tested?
- Update the ocw-course starter config with the [new line here](https://github.com/mitodl/ocw-hugo-projects/pull/180/files).
- Start studio with concourse `docker-compose --profile concourse up`
- Run `manage.py upsert_mass_build_pipeline` and `backpopulate_pipelines`, they should both succeed.
- Try to publish an existing site with a `publish_date`, the drawer should look about the same and work the same.
- Try to publish a new or existing site without a `publish_date`, you should see a URL field with a suggested path, any existing metadata values matching the pattern should be present.  Enter a valid value and click 'Publish' to draft, check to see that the `url_path` is updated and the pipeline is triggered (it will fail, that's ok).
- Go to the github repo and check that any resources for images, pdfs, etc have a `file` value that contains the site's url path.
- Reload the page and try to publish again, the URL field should be set to whatever you entered last time instead of the suggested path.  Modify it and click the "Publish" button, the pipeline should trigger again.  Also check on the git repo and make sure the resource file paths have been updated again to the latest value.
- The changes here affect most parts of the site, so try just about everything and make sure it still works.
- The changes to the pipelines are best tested when this hits RC.

#### Where should the reviewer start?
https://github.com/mitodl/hq/discussions/1

#### Screenshots (if appropriate)
An unpublished site without any user-supplied url path so far:

Screen Shot 2022-05-17 at 3 45 16 PM

An unpublished site with a previously chosen url (still editable):

Screen Shot 2022-05-17 at 3 53 26 PM

A published site (no longer editable):

Screen Shot 2022-05-17 at 3 54 10 PM

### CODE REVIEW

File: websites/serializers.py [hunk 1]
```diff
@@ -519,7 +567,7 @@ class Meta:
 class WebsiteContentCreateSerializer(
     serializers.ModelSerializer, RequestUserSerializerMixin
 ):
-    """Serializer which creates a new WebsiteContent"""
+    """Serializer mixin which validates that urls are unique"""
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply websites/serializers.py:570]: Nope, pasted in the wrong place

File: static/js/components/forms/PublishForm.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+export const websiteUrlValidation = yup.object().shape({
+  url_path: yup
+    .string()
+    .label("URL Path")
+    .trim()
+    .required()
+    .matches(
+      /^[a-zA-Z0-9\-._]*$/,
```
[the remark this answers is not in the document]
Comment (Ferdi) [reply static/js/components/forms/PublishForm.tsx:31]: I personally don't like underscores in the url but I dont think we should disallow them .

**Comment by ChristopherChudzicki** [comment]
@mbertrand I tried the pipelines / renaming etc and everything worked as expected: URLs got updated in github when I changed the course url 👍

File: websites/models.py [hunk 3]
```diff
@@ -172,13 +194,73 @@ def get_url(self, version=VERSION_LIVE):
...
+        if self.starter is None:
+            return None
+        site_config = SiteConfig(self.starter.config)
+        url_format = site_config.site_url_format
+        if not url_format or self.publish_date:
+            # use name for published  sites or for any sites without a `url_path` in config.
+            url_format = self.name
+        elif url_format:
+            for section in re.findall(r"(\[.+?\])+", site_config.site_url_format) or []:
+                section_type, section_field = re.sub(r"[\[\]]+", "", section).split(":")
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply websites/models.py:239]: Nice idea!  But it doesn't quite work as intended if any of the metadata fields haven't been filled out yet.  For example, when I try it on a site with fields for `primary_course_number`  and `title` but no `term` and `year`, I get:

```python
slugify(site_config.site_url_format.format(**metadata).replace(".", "-"))

>2.18-time-travel-101-abc-
```

But what I want is placeholders for the missing values like `2.18-time-travel-101-abc-{term}-{year}`

**Comment by mbertrand** [comment]
> Do you think the frontend form using outdated metadata is a blocking issue? Probably not?

I think I know how to fix that, will give it a shot.

**Comment by gumaerc** [comment]
I'm not sure I could point out anything from a technical standpoint that @ChristopherChudzicki hasn't already mentioned, but for what it's worth I experienced the same behavior where updating my metadata wasn't reflected on the URL in the publish drawer until I refreshed the page.  One UI thing that I noticed that I didn't see pointed out is that with the new text field in the publish drawer, it seems to open super wide:

image

Do we maybe want to restrict the width on this?  I tried publishing my test course, which already existed, and I noticed that the `file` attributes have a seemingly random amount of `/media` prefixed at the beginning.  Is this maybe something to do with my local env, or the test course itself?  You can take a look at the resources in my published test course here: https://github.com/gumaerc-testorg/cpg-test-course-3/tree/main/content/resources

**Comment by mbertrand** [comment]
I made the publish drawer deliberately wide if the site hasn't been published to production yet, because the `site-url-format` string that is the default value in the URL field is so long.  I mentioned to Ferdi that maybe we should come up with some shorter value to display, maybe without the `sitemetadata:` in each section, so `[primary_course_number]-[title]-[term]-[year]` (though that's still a bit long).

If you're seeing `/media` for files. you probably don't have `OCW_STUDIO_USE_S3` set to `True` in your .env file.

**Comment by mbertrand** [comment]
The suggested url format in the publish drawer should now be updated after a metadata change.

**Review by ChristopherChudzicki — APPROVED** [review]

**Review by gumaerc — APPROVED** [review]
:+1:

**Pull request merged**